runtime/src/main/java/org/apache/flink/agents/runtime/operator/ActionExecutionOperator.java

```diff
@@ -0,0 +1,271 @@
+        this.runnerContext = new RunnerContextImpl();
+        this.chainingStrategy = ChainingStrategy.ALWAYS;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+
+        reusedStreamRecord = new StreamRecord<>(null);
+        reusedSideOutputStreamRecord = new StreamRecord<>(null);
+
+        // init state
+        ValueStateDescriptor<Integer> pendingActionCountDescriptor =
+                new ValueStateDescriptor<>("pendingActionCount", TypeInformation.of(Integer.class));
+        pendingActionCountState = getRuntimeContext().getState(pendingActionCountDescriptor);
+        ListStateDescriptor<EventMessage<K>> pendingInputEventsDescriptor =
+                new ListStateDescriptor<>("pendingInputEvents", eventMessageTypeInfo);
+        pendingInputEventsState = getRuntimeContext().getListState(pendingInputEventsDescriptor);
+
+        // init PythonActionExecutor
+        initPythonActionExecutor();
+    }
+
+    @Override
+    public void processElement(StreamRecord<EventMessage<K>> record) throws Exception {
+        EventMessage<K> eventMessage = record.getValue();
+        LOG.debug("ActionExecutionOperator receive element {}", eventMessage);
+
+        if (eventMessage.isInputEvent() && (hasPendingAction() || hasPendingInputEvent())) {
+            // If the event is an InputEvent and there are pending actions or input events with the
+            // same key, it will be added to pendingInputEvents and processed later.
+            LOG.debug("Add element {} to pending list.", eventMessage);
+            pendingInputEventsState.add(eventMessage);
+            return;
+        }
+
+        processEventMessage(eventMessage);
+        tryProcessPendingInputEvents();
+    }
+
+    private void processEventMessage(EventMessage<K> eventMessage) throws Exception {
+        LOG.debug("ActionExecutionOperator process element {}", eventMessage);
+        Event event = eventMessage.getEvent();
+        List<Action> actions = workflowPlan.getEventTriggerActions(event.getClass());
+        if (actions != null && !actions.isEmpty()) {
+            // add pending action count for InputEvent.
+            // We should record the number of actions to be processed for each event only once.
+            // For InputEvents that must originate from the regular input stream, we increment the
+            // action count once before processing the event.
+            // For other Events, which are guaranteed to be generated by an action and enter the
+            // feedback loop, we increment the action count once at generation time.
+            if (eventMessage.isInputEvent()) {
+                addPendingActionCount(actions.size());
+            }
+            for (Action action : actions) {
+                // execute action
+                List<Event> actionOutputEvents;
+                if (action.getExec() instanceof JavaFunction) {
+                    action.getExec().call(event, runnerContext);
```

Comment:
Execution of action can take long, thus should be asynchronous. Let's mark a todo here.